mqtt/mqtt-broker/src/auth/authorization.rs

```diff
@@ -0,0 +1,217 @@
+use mqtt3::proto;
+
+use crate::{AuthId, ClientId, Error};
+
+/// A trait to check a MQTT client permissions to perform some actions.
+#[async_trait]
+pub trait Authorizer {
+    /// Authorizes a MQTT client to perform some action.
+    async fn authorize(&self, activity: Activity) -> Result<bool, Error>;
+}
+
+#[async_trait]
+impl<F> Authorizer for F
+where
+    F: Fn(Activity) -> Result<bool, Error> + Sync,
+{
+    async fn authorize(&self, activity: Activity) -> Result<bool, Error> {
+        self(activity)
+    }
+}
+
+/// Default implementation that always denies any operation a client intends to perform.
+/// This implementation will be used if custom authorization mechanism was not provided.
+pub struct DefaultAuthorizer;
+
+#[async_trait]
+impl Authorizer for DefaultAuthorizer {
+    async fn authorize(&self, _: Activity) -> Result<bool, Error> {
+        Ok(false)
+    }
+}
+
+/// Describes a client activity to authorized.
+pub struct Activity {
+    auth_id: AuthId,
+    client_id: ClientId,
+    operation: Operation,
+}
+
+impl Activity {
+    pub fn new(
+        auth_id: impl Into<AuthId>,
+        client_id: impl Into<ClientId>,
+        operation: Operation,
+    ) -> Self {
+        Self {
+            auth_id: auth_id.into(),
+            client_id: client_id.into(),
+            operation,
+        }
+    }
+
+    pub fn operation(&self) -> &Operation {
+        &self.operation
+    }
+}
+
+/// Describes a client operation to be authorized.
+pub enum Operation {
+    Connect(Connect),
+    Publish(Publish),
+    Subscribe(Subscribe),
+    Receive(Receive),
+}
+
+impl Operation {
+    /// Creates a new operation context for CONNECT request.
+    pub fn new_connect(connect: proto::Connect) -> Self {
+        Self::Connect(connect.into())
+    }
+
+    /// Creates a new operation context for PUBLISH request.
+    pub fn new_publish(publish: proto::Publish) -> Self {
+        Self::Publish(publish.into())
+    }
+
+    /// Creates a new operation context for SUBSCRIBE request.
+    pub fn new_subscribe(subcribe: proto::Subscribe) -> Self {
+        Self::Subscribe(subcribe.into())
+    }
+
+    /// Creates a new operation context for RECEIVE request.
+    ///
+    /// RECEIVE request happens when broker decides to publish a message to a certain
+    /// topic client subscribed to.
+    pub fn new_receive(publication: proto::Publication) -> Self {
+        Self::Receive(publication.into())
+    }
+}
+
+/// Represents a client attempt to connect to the broker.
+pub struct Connect {
+    will: Option<Publication>,
+}
+
+impl From<proto::Connect> for Connect {
+    fn from(connect: proto::Connect) -> Self {
+        Self {
+            will: connect.will.map(Into::into),
+        }
+    }
+}
+
+/// Represents a publication description without payload to be used for authorization.
+pub struct Publication {
+    pub topic_name: String,
+    pub qos: proto::QoS,
+    pub retain: bool,
+}
+
+impl From<proto::Publication> for Publication {
```

Comment:
You won't be able to use the From interface anymore, but it might be better to create this publication from a reference of a proto::publication to avoid having to clone the payload. 

(I know cloning bytes only increments the ref count, but its still extra overhead)

Reply:
Not sure if will give us too much benefit... Comparing to serializing data to send to edgehub-authorizer it may cost almost nothing.